### PR TITLE
[ALTS] Create GsecKeyInterface and related classes in ALTS

### PR DESCRIPTION
**Pull request by rockspore**
As such, `alts_zero_copy_grpc_protector_create` will take a `GsecKeyFactoryInterface` to create `GsecKeyInterface` objects for the underlying crypter to use.

This enables the caller to control how all the key related buffers are prepared and protected.

`gsec_aes_gcm_aead_crypter` holds the raw pointer to `GsecKeyInterface` instead of a `unique_ptr` possibly because somewhere in the test (and maybe production code as well), the structure is getting copied. A SEGV error would be caused with `unique_ptr` which doesn't support copy operations.

### CODE REVIEW

File: test/core/tsi/alts/crypt/aes_gcm_test.cc [hunk 1]
```diff
@@ -755,8 +758,10 @@ static void gsec_test_get_crypter_from_test_vector(
   size_t ciphertext_and_tag_length = test_vector->ciphertext_and_tag_length;
   ASSERT_EQ(ciphertext_and_tag_length, plaintext_length + kAesGcmTagLength);
   size_t tag_length = ciphertext_and_tag_length - plaintext_length;
-  gsec_aes_gcm_aead_crypter_create(test_vector->key, key_length, nonce_length,
-                                   tag_length, rekey, crypter, nullptr);
+  gsec_aes_gcm_aead_crypter_create(
+      std::make_unique<grpc_core::GsecKey>(
+          absl::MakeConstSpan(test_vector->key, key_length), rekey),
+      nonce_length, tag_length, crypter, nullptr);
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/crypt/aes_gcm_test.cc:764]: micro-nit: Can you add `/*error_details=*/` in front of `nullptr`?

File: test/core/tsi/alts/crypt/aes_gcm_test.cc [hunk 2]
```diff
@@ -803,8 +808,10 @@ static void gsec_test_create_random_aes_gcm_crypter(gsec_aead_crypter** crypter,
                                                     bool rekey) {
   uint8_t* key;
   gsec_test_random_array(&key, key_length);
-  gsec_aes_gcm_aead_crypter_create(key, key_length, nonce_length, tag_length,
-                                   rekey, crypter, nullptr);
+  gsec_aes_gcm_aead_crypter_create(
+      std::make_unique<grpc_core::GsecKey>(absl::MakeConstSpan(key, key_length),
+                                           rekey),
+      nonce_length, tag_length, crypter, nullptr);
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/crypt/aes_gcm_test.cc:814]: micro-nit: Can you add `/*error_details=*/` in front of `nullptr`?

File: test/core/tsi/alts/frame_protector/alts_crypter_test.cc [hunk 3]
```diff
@@ -354,28 +357,46 @@ static void create_random_alts_seal_crypter(
...
-                                   kAesGcmTagLength, rekey, client_crypter_seal,
-                                   nullptr);
-  gsec_aes_gcm_aead_crypter_create(key, key_length, kAesGcmNonceLength,
-                                   kAesGcmTagLength, rekey,
-                                   client_crypter_unseal, nullptr);
+  ASSERT_EQ(
+      gsec_aes_gcm_aead_crypter_create(
+          std::make_unique<grpc_core::GsecKey>(
+              absl::MakeConstSpan(key, key_length), rekey),
+          kAesGcmNonceLength, kAesGcmTagLength, server_crypter_seal, nullptr),
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/frame_protector/alts_crypter_test.cc:364]: micro-nit: Can you add `/*error_details=*/` in front of `nullptr`?

Similarly below

File: test/core/tsi/alts/zero_copy_frame_protector/alts_grpc_record_protocol_test.cc [hunk 4]
```diff
@@ -124,35 +126,39 @@ test_fixture_integrity_only_create(bool rekey, bool extra_copy) {
 
   // Create client record protocol for protect.
   EXPECT_EQ(gsec_aes_gcm_aead_crypter_create(
-                key, key_length, kAesGcmNonceLength, kAesGcmTagLength, rekey,
-                &crypter, nullptr),
+                std::make_unique<grpc_core::GsecKey>(
+                    absl::MakeConstSpan(key, key_length), rekey),
+                kAesGcmNonceLength, kAesGcmTagLength, &crypter, nullptr),
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_grpc_record_protocol_test.cc:131]: micro-nit: Can you add `/*error_details=*/` in front of `nullptr`?

Similarly below.

File: test/core/tsi/alts/zero_copy_frame_protector/alts_iovec_record_protocol_test.cc [hunk 5]
```diff
@@ -106,35 +109,39 @@ alts_iovec_record_protocol_test_fixture_create(bool rekey,
   gsec_aead_crypter* crypter = nullptr;
   // Create client record protocol for protect.
   EXPECT_EQ(gsec_aes_gcm_aead_crypter_create(
-                key, key_length, kAesGcmNonceLength, kAesGcmTagLength, rekey,
-                &crypter, nullptr),
+                std::make_unique<grpc_core::GsecKey>(
+                    absl::MakeConstSpan(key, key_length), rekey),
+                kAesGcmNonceLength, kAesGcmTagLength, &crypter, nullptr),
```
Comment (matthewstevenson88) [inline test/core/tsi/alts/zero_copy_frame_protector/alts_iovec_record_protocol_test.cc:114]: micro-nit: Can you add `/*error_details=*/` in front of `nullptr`?

Similarly below.

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc [hunk 6]
```diff
@@ -119,9 +118,10 @@ static tsi_result create_alts_grpc_record_protocol(
   grpc_status_code status;
   gsec_aead_crypter* crypter = nullptr;
   char* error_details = nullptr;
-  status = gsec_aes_gcm_aead_crypter_create(key, key_size, kAesGcmNonceLength,
-                                            kAesGcmTagLength, is_rekey,
-                                            &crypter, &error_details);
+  bool is_rekey = key->IsRekey();
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc:121]: Why is this line needed?

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 7]
```diff
@@ -193,8 +190,11 @@ static tsi_result handshaker_result_create_zero_copy_grpc_protector(
           "protector equals %zu",
           *max_output_protected_frame_size);
   tsi_result ok = alts_zero_copy_grpc_protector_create(
-      reinterpret_cast<const uint8_t*>(result->key_data),
-      kAltsAes128GcmRekeyKeyLength, /*is_rekey=*/true, result->is_client,
+      grpc_core::GsecKeyFactory(
+          absl::MakeConstSpan(reinterpret_cast<uint8_t*>(result->key_data),
+                              kAltsAes128GcmRekeyKeyLength),
+          true),
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc:196]: micro-nit: Please add `/*is_rekey=*/` in front of `true`.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 8]
```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
...
+GsecKey::GsecKey(absl::Span<const uint8_t> key, bool is_rekey)
+    : is_rekey_(is_rekey) {
+  if (is_rekey_) {
+    aead_key_ = std::vector<uint8_t>(kRekeyAeadKeyLen);
+    kdf_buffer_ = std::vector<uint8_t>(EVP_MAX_MD_SIZE);
+    nonce_mask_ = std::vector<uint8_t>(kAesGcmNonceLength);
+    memcpy(nonce_mask_.data(), key.data() + kKdfKeyLen, kAesGcmNonceLength);
+    kdf_counter_ = std::vector<uint8_t>(kKdfCounterLen, 0);
+  }
+  key_ = std::vector<uint8_t>(is_rekey_ ? kKdfKeyLen : key.size());
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/crypt/aes_gcm.cc:60]: nit: I think you can use `key_.resize(...)` instead of creating a new vector.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 9]
```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
...
+  memcpy(key_.data(), key.data(), key_.size());
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/crypt/aes_gcm.cc:61]: Is this line correct? Should it be `is_rekey_ ? kKdfKeyLen : key.size()` instead?

If yes, can we add a unit test that would have caught this?

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 10]
```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
...
+}
+
+bool GsecKey::IsRekey() { return is_rekey_; }
+
+absl::Span<const uint8_t> GsecKey::key() { return absl::MakeConstSpan(key_); }
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/crypt/aes_gcm.cc:66]: What's the rationale for using MakeConstSpan for some of these APIs and MakeSpan for others?

**Review by matthewstevenson88 — COMMENTED** [review]
Thanks for the PR Luwei!

File: test/core/tsi/alts/crypt/aes_gcm_test.cc (hunk 2 above)
Comment (rockspore) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: Done.

File: test/core/tsi/alts/crypt/aes_gcm_test.cc (hunk 1 above)
Comment (rockspore) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: Done.

File: src/core/tsi/alts/crypt/aes_gcm.cc (hunk 9 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc:61]: Ah of course! Sorry I missed this. :)

File: src/core/tsi/alts/crypt/aes_gcm.cc (hunk 10 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc:66]: I think it's fine to leave it as-is: what I was missing before was I did not notice the difference in the return types (i.e. one returns a `uint8_t` span versus the others return a `const uint8_t` span).

File: src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc (hunk 6 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc:121]: Makes sense, thanks for clarifying!

**Comment by matthewstevenson88** [comment]
Re-running kokoro just to ensure that the grpclb e2e test flake is not related to this PR.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 11]
```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
 constexpr size_t kKdfCounterOffset = 2;
 constexpr size_t kRekeyAeadKeyLen = kAes128GcmKeyLength;
 
+namespace grpc_core {
+
+GsecKeyFactory::GsecKeyFactory(absl::Span<const uint8_t> key, bool is_rekey)
+    : key_(std::vector<uint8_t>(key.begin(), key.end())), is_rekey_(is_rekey) {}
```
Comment (matthewstevenson88) [inline src/core/tsi/alts/crypt/aes_gcm.cc:45]: micro-nit: Do we need the `std::vector<uint8_t>` here? Is it sufficient to write
```
: key_(key.begin(), key.end())
```
instead?

Comment (rockspore) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: Done. In fact, I also removed `absl::MakeConstSpan` in some (unfortunately not all) places, as the compiler is smart enough to recognize `{key, key_length}` initialization.

**Comment by rockspore** [comment]
> Re-running kokoro just to ensure that the grpclb e2e test flake is not related to this PR.

I think those failing tests passed.

**Review by matthewstevenson88 — APPROVED** [review]
